lib/spack/llnl/util/tty/log.py

```diff
@@ -813,12 +813,16 @@ def background_reader(reader, echo_writer, _kill):
                     is_killed = _kill.wait(.1)
                     self.stderr.flush()
                     self.stdout.flush()
-                    line = reader.readline()
-                    while line:
+
+                    while not reader.closed:
                         if self.echo:
+                            try:
+                                line = reader.readline()
```

Comment:
For files in general `readline` returns `''` to indicate it is done, so I would think the original code would work. Does something go wrong if the changes in this function are reverted?

Reply:
Its more an issue of `readline` attempting to read from a closed fd. If these changes are reverted, there is an exception in each log stating as much.

Reply:
That makes sense but I'm curious: if we close the `reader` in `__exit__` could it truncate output? If we close the writer, will the reader eventually see `EOF` (in that case the reader thread could be responsible for closing the file)? This doesn't seem like a major issue since the "missing" output would still be in the log file but potentially worth considering.